mypy/dmypy.py

```diff
@@ -0,0 +1,585 @@
+                    except OSError as err:
+                        pass  # Maybe the client hung up
+                    conn.close()
+                    if command == 'stop':
+                        sock.close()
+                        sys.exit(0)
+            finally:
+                os.unlink(STATUS_FILE)
+        finally:
+            os.unlink(self.sockname)
+
+    def create_listening_socket(self) -> socket.socket:
+        """Create the socket and set it up for listening."""
+        self.sockname = os.path.abspath(SOCKET_NAME)
+        if os.path.exists(self.sockname):
+            os.unlink(self.sockname)
+        sock = socket.socket(socket.AF_UNIX)
+        sock.bind(self.sockname)
+        sock.listen(1)
+        return sock
+
+    def run_command(self, command: str, data: Mapping[str, object]) -> Dict[str, object]:
+        """Run a specific command from the registry."""
+        key = 'cmd_' + command
+        method = getattr(self.__class__, key, None)
+        if method is None:
+            return {'error': "Unrecognized command '%s'" % command}
+        else:
+            return method(self, **data)
+
+    # Command functions (run in the server via RPC).
+
+    def cmd_status(self) -> Dict[str, object]:
+        """Return daemon status."""
+        res = {}  # type: Dict[str, object]
+        res.update(get_meminfo())
+        return res
+
+    def cmd_stop(self) -> Dict[str, object]:
+        """Stop daemon."""
+        return {}
+
+    last_sources = None
+
+    def cmd_check(self, files: Sequence[str]) -> Dict[str, object]:
+        """Check a list of files."""
+        # TODO: Move this into check(), in case one of the args is a directory.
+        # Capture stdout/stderr and catch SystemExit while processing the source list.
+        save_stdout = sys.stdout
+        save_stderr = sys.stderr
+        try:
+            sys.stdout = stdout = io.StringIO()
+            sys.stderr = stderr = io.StringIO()
+            self.last_sources = mypy.main.create_source_list(files, self.options)
+        except SystemExit as err:
+            return {'out': stdout.getvalue(), 'err': stderr.getvalue(), 'status': err.code}
+        finally:
+            sys.stdout = save_stdout
+            sys.stderr = save_stderr
+        return self.check(self.last_sources)
+
+    def cmd_recheck(self) -> Dict[str, object]:
+        """Check the same list of files we checked most recently."""
+        if not self.last_sources:
+            return {'error': "Command 'recheck' is only valid after a 'check' command"}
+        return self.check(self.last_sources)
+
+    last_mananager = None  # type: Optional[mypy.build.BuildManager]
+
+    def check(self, sources: List[mypy.build.BuildSource],
+              alt_lib_path: Optional[str] = None) -> Dict[str, Any]:
+        bound_gc_callback = lambda phase, info: self.gc_callback(phase, info)
+        self.gc_start_time = None  # type: Optional[float]
```

Comment:
I'd recommend moving the stats handling code to separate methods to keep the logic of this method less cluttered.